src/mlpack/methods/bayesian_ridge/bayesian_ridge.cpp

```diff
@@ -0,0 +1,299 @@
+/**
+ * @file bayesian_ridge.cpp
+ * @author Clement Mercier 
+ *
+ * Implementation of Bayesian Ridge regression.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include "bayesian_ridge.hpp"
+#include <mlpack/core/util/log.hpp>
+#include <mlpack/core/util/timers.hpp>
+
+using namespace mlpack;
+using namespace mlpack::regression;
+
+
+BayesianRidge::BayesianRidge(const bool fitIntercept,
+                             const bool normalize) :
+  fitIntercept(fitIntercept),
+  normalize(normalize)
+{/* Nothing to do */}
+
+float BayesianRidge::Train(const arma::mat& data,
+                           const arma::rowvec& responses)
+{
+  Timer::Start("bayesian_ridge_regression");
+
+  arma::mat phi;
+  arma::rowvec t;
+  arma::colvec vecphitT;
+  arma::mat phiphiT;
+  arma::colvec eigval;
+  arma::mat eigvec;
+  arma::colvec eigvali;
+
+  // Preprocess the data. Center and normalize.
+  CenterNormalize(data,
+                  responses,
+                  fitIntercept,
+                  normalize,
+                  phi,
+                  t,
+                  data_offset,
+                  data_scale,
+                  responses_offset);
+  vecphitT = phi * t.t();
+  phiphiT =  phi * phi.t();
+
+  arma::eig_sym(eigval, eigvec, phiphiT);
+
+  // Detect singular matrix.
+  if (eigval[0] == 0)
+  {
+    Log::Warn << "Singular matrix. Two lines or more are colinear."
+              <<  std::endl;
+    return -1;
+  }
+
+  unsigned short p = data.n_rows, n = data.n_cols;
+  // Initialize the hyperparameters and
+  // begin with an infinitely broad prior.
+  alpha = 1e-6;
+  beta =  1 / (var(t) * 0.1);
+
+  double tol = 1e-3;
+  unsigned short nIterMax = 50;
+  unsigned short i = 0;
+  double deltaAlpha = 1, deltaBeta = 1, crit = 1;
+  arma::mat matA = arma::eye<arma::mat>(p, p);
+  arma::rowvec temp;
+
+  while ((crit > tol) && (i < nIterMax))
+    {
+      deltaAlpha = -alpha;
+      deltaBeta = -beta;
+
+      // Compute the posterior statistics.
+      // with inv()
+      for (size_t k = 0; k < p; k++) {matA(k, k) = alpha;}
+      // inv is used instead of solve beacause we need matCovariance to
+      // compute the prediction uncertainties. If solve is used, matCovariance
+      // must be comptuted at the end of the loop.
+      matCovariance = inv_sympd(matA + phiphiT * beta);
+      omega = (matCovariance * vecphitT) * beta;
```

Comment:
We could inline `matCovariance` here as it's only used once.

Reply:
I was not sure f what to do exactly. Do I have to declare matCovariance inline in the .hpp ?

Reply:
Sorry for the confusion, I meant to write:

`mega = arma::nv_sympd(matA + phiphiT * beta) * vecphitT * beta;`.

Reply:
The covariance matrix is stored because it is required to compute the predictive posterior uncertainties in the Predict() method. However if you think that it is not relevant the compute the uncertainties for this estimator, we can suppress the associated Predict() method and inline matCovariance during the training phase.

Reply:
I think that all your previous comments have been taken into account in the last modifications.

Reply:
Thanks for the clarification, I see, I missed that. In this case let's keep it as is.